mplc/scenario.py

```diff
@@ -550,19 +562,196 @@ def split_data_advanced(self, is_logging_enabled=True):
             p.x_train = np.concatenate(list_arrays_x)
             p.y_train = np.concatenate(list_arrays_y)
 
-            # Create local validation and test datasets from the partner train data
-            p.x_train, p.x_val, p.y_train, p.y_val = train_test_split(
-                p.x_train, p.y_train, test_size=0.1, random_state=42
-            )
-            p.x_train, p.x_test, p.y_train, p.y_test = train_test_split(
-                p.x_train, p.y_train, test_size=0.1, random_state=42
-            )
-
         # Check coherence of number of mini-batches versus partner with small dataset
         assert self.minibatch_count <= min(
             [len(p.x_train) for p in self.partners_list]
         ), "Error: in the provided \
-            config file and the provided dataset, a partner doesn't have enough data samples to create the minibatches "
+                    config file and the provided dataset," \
+           " a partner doesn't have enough data samples to create the minibatches "
+
+        if self.val_set == 'local':
+            # Advanced split: Populates the partners with their val data
+
+            y_val = LabelEncoder().fit_transform([str(y) for y in self.dataset.y_val])
+
+            # Stratify the dataset into clusters per labels
+            x_val_for_cluster, y_val_for_cluster, nb_samples_per_cluster_val = {}, {}, {}
+            for label in labels:
+                idx_in_full_valset = np.where(y_val == label)
+                x_val_for_cluster[label] = self.dataset.x_val[idx_in_full_valset]
+                y_val_for_cluster[label] = self.dataset.y_val[idx_in_full_valset]
+                nb_samples_per_cluster_val[label] = len(y_val_for_cluster[label])
```

Comment:
This leads to a lot lot of code duplication. Couldn't we factorize `split_data_advanced` in a function which we would apply to the datasets? Or sequence it the other way around, first the split among partners, then the val and test split?
Not sure what would be best, but we should try to avoid duplicating too much code.

Reply:
Done with splitter object !